File: competency-management-frontend/src/app/_services/user.service.ts
import {Injectable} from "@angular/core";
import {HttpClient} from "@angular/common/http";
import {User} from "../_model/user.model";
import {Skill} from "../_model/skill.model";

@Injectable({
  providedIn: 'root'
})
export class UserService {
  private static USERS_ENDPOINT = "http://localhost:8080/users";

  constructor(private http: HttpClient) {}

  fetchAll() {
    return this.http.get<User[]>(UserService.USERS_ENDPOINT + '/all');
  }

  fetchById(userId: number) {
    return this.http.get<User>(`${UserService.USERS_ENDPOINT}/${userId}`);
  }

  fetchLoggedInUser() {
    return this.http.get<User>(UserService.USERS_ENDPOINT);
  }

  updateUserSkills(userId: number, skillLevelIds: number[]) {
    return this.http.post<User>(`${UserService.USERS_ENDPOINT}/${userId}/update-skills`, skillLevelIds);
  }

  saveAndAssignUserSkill(userId: number, skill: Skill) {
    return this.http.post<User>(`${UserService.USERS_ENDPOINT}/${userId}/assign-skill`, skill);
  }

  deleteUserSkill(userSkillId: number) {
    return this.http.delete<number>(`${UserService.USERS_ENDPOINT}/${userSkillId}`);
  }
}

File: competency-management-backend/src/main/java/org/example/competencymanagementsystem/service/UserService.java
package org.example.competencymanagementsystem.service;

import org.example.competencymanagementsystem.model.dto.*;

import java.util.List;

public interface UserService {
    List<UserDto> fetchAllAuthenticated();
    UserDto fetchById(Long userId);
    UserDto login(CredentialsDto credentialsDto);
    UserDto register(SignUpDto signUpDto);
    UserDto fetchLoggedInUser();
    UserDto updateUserSkills(Long userId, List<Long> skills);
    UserDto saveAndAssignUserSkill(Long userId, SkillDTO skillDTO);
    Long deleteUserSkill(Long userSkillId);
}
